### PR TITLE
Fix loss scaling

### LINKED ISSUE #30: Loss calculation issue
```
            avg_cost = sum(cost_values) / len(batch)
        aggregated_train_loss += avg_cost
    train_loss = np.sqrt(aggregated_train_loss / len(train_ds))
    logging.info(f"RMS train loss: {train_loss}")
    train_losses.append(train_loss)
```

We already compute `avg_cost = sum(cost_values) / len(batch) ` for each batch.

So we should divide by number of batches instead of `len(train_ds)`

Or we change `avg_cost = sum(cost_values)`

```
            avg_cost = sum(cost_values)
        aggregated_train_loss += avg_cost
    train_loss = np.sqrt(aggregated_train_loss / len(train_ds))
    logging.info(f"RMS train loss: {train_loss}")
    train_losses.append(train_loss)
```

And `len(batch)` is not the number of data points. `len(batch["symbols"])` represents that.

### LINKED ISSUE #31: Scaling issue of down sampling
Our down sampling occurs scaling issue. Look at the y-axis scale of the figures below. 

down sampling by 2**6
Image

down sampling by 2*14
Image

Let me explain why we have scaling mismatch issue. 
For example, if the original n_grid is 102, we're gonna downsample it to 34.

Without downsampling,
```
coefficients = [c1, c2, c3, c4, ..., c100, c101, c102]
densities= [d1, d2, d3, d4, ... ,d100, d101, d102]
grid = [w1, w2, ... ,w100, w101, w102]
xc_energy_density = [c1 d1, c2 d2, c3 d3, ... ,c100 d100, c101 d101, c102 d102]
xc_energy = a1 b1 w1 + ... + a102 b102 w102
```

With downsampling,
```
coefficients_down = [c1 + c2 + c3, c4 + c5 + c6 , ... , c100 + c101 + c102]
#QNNFunctional.compute_slice_sums(unscaled_densities, indices)
densities_down = [d1 + d2 + d3, d4 + d5 + d6, ..., d100 + d101 + d102] 
grid_down = [w1 + w2 + w3, ... , w100 + w101 + w102]
xc_energy_density_down = [(c1+ c2+ c3)(d1+d2+d3), ... , (c100+c101+c102)(d100+d101+d102)]
xc_energy_down = (c1+ c2+ c3)(d1+d2+d3)(w1+w2+w3) + ... + (c100+c101+c102)(d100+d101+d102)(w100+w101+w102)
```

let's suppose K is number of new indicies. (For above example, $$K =34, |G_k|=3$$). We can mathematically write those as
$$E_{xc}= \sum_{i=1}^{K \ G_k} c_i d_i w_i$$
$$E_{xc, down} =\sum_{k=1}^K (\sum_{i \in G_k} c_i)(\sum_{i \in G_k} d_i)(\sum_{i \in G_k} w_i)$$.

$$E_{xc, down}$$ scale is bigger as $K$  decrease or $|G_k|$ increase. 
I argue that we need to divide it by $K^\alpha$. I think $K^2$ works well.

Modified code will be like below. 
```
xc_energy_density = jnp.einsum("rf,rf->r", coefficients, densities)
xc_energy_density = abs_clip(xc_energy_density, clip_cte)
xc_energy_down= self._integrate(xc_energy_density, grid_weights)#sungwon
new_indicies_size = len(grid.weights)//(2**n_qubits)
return xc_energy_down/new_indicies_size**2
```

### PR DESCRIPTION
**Pull request by minhtriet**
Fix #30 @wonagram

### CODE REVIEW

**Comment by wonagram** [comment]
I still don't understand ... 

```
len(train_ds)=29
len(range(0, len(train_ds), batch_size)))=10 #number of batches
Batch1 = [c1, c2, c3]
Batch2 = [c4, c5, c6]
...
Batch 9 = [c25, c26, c27]
Batch 10 =[c28, c29]
```

This is what you’ve done
((c1+c2+c3)/3 + ... +(c25+c26+c27)/3+ (c28+c29)/2) /10 = (c1+c2+c3)/30 + (c25+c26+c27)/30 + (c28+c29)/20

but this is what I’m trying to do.
(c1+c2+ ... c29)/29

**Comment by minhtriet** [comment]
I think we are having two goals here. 
- Your goal is to calculate the correct `train_loss` for the report.
- My goal is to do mini batch gradient descent, which means we divide the training set into small baches and evaluate the gradient based on a batch, update the parameter, and learn using the next batch.

I agree with your math. We can drop the last batch with 2 elements then the math from my method should be equal to yours. These dropped elements will be fed to the QNN in later epochs because we shuffle the training set

What do you think? Thank you for being thorough

**Review by wonagram — APPROVED** [review]
